### ISSUE TITLE
Add `util.scope` op/interface.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#16081]
For awhile we've needed an op (or set of ops) that lets us attribute regions of ops nested within other ops. This would let us specify device targets for a particular grouping of ops, annotate debugging information/tracing zones, or author outlining passes (such as for reusable command buffers) that act incrementally (like dispatch region formation). `scf.execute_region` is kind of like that but insufficient: we need to be able to support tied operands, captured dynamic shapes/sizes, and to have transformations be able to mutate the types as we convert between dialects/stages.

The plan is to fork `scf.execute_region` into `util.scope` and support the various op interfaces we have for util types.
The op should not have any change on execution behavior and always be safe to insert and remove. We'll need patterns to inducing tying of operands and awareness of the op in passes that perform structural walks (DFX, ScheduleAllocation, ScheduleExecution, etc). We could do this with an op interface so that we can have dialect-specific ops/syntactic sugar - like a `hal.device.scope` or something - but with external interfaces we could also make a single `util.scope` op implement dialect-specific interfaces. It'd be a tradeoff of needing to do some deep nesting to add 3 bits of information `dialect_a.scope { dialect_b.scope { dialect_c.scope { actual stuff } } }` vs `util.scope attrs { actual stuff }`. It's probably be useful to have some dedicated tracing ops (for non-structured zone begin/end, etc) but a util.scope with an annotation could be lowered into those ops if they existed. 

Examples:
```mlir
// empty (or non-returning)
util.scope {
  util.yield
}

// implicit capture
%c0 = arith.constant 0 : i32
%0 = util.scope -> i32 {
  util.yield %c0 : i32
}

// shape/size-aware types
%0 = util.scope -> (tensor<4x?xf32>{%dim}, !stream.resource<*>{%size}) {
  util.yield ...
}

// tied operands
%0 = util.scope -> (%tied as tensor<4xf32>{%dim}) {
  util.yield %tied : tensor<4xf32>{%dim}
}

// optional attrs for `hal.device.targets`/`stream.affinity`/etc
%0 = util.scope {
  hal.device.targets = [ ... ]
} -> i32 {
  // ops on device A
  util.yield ...
}
%1 = util.scope {
  hal.device.targets = [ ... ]
} -> i32 {
  // ops on device B
  util.yield ...
}

// optional naming we could turn into trace zones (informational only)
%0 = util.scope "some name" -> i32 { ... }
// or if we had an interface:
%0 = util.trace.scope "some name" -> i32 { ... }

// scopes that end up entirely on-device can get device trace zones
%0 = stream.async.execute ... {
  %1 = util.scope "first part" {
    %t0 = stream.async.dispatch ...
    %t1 = stream.async.dispatch ...
    util.yield %t1
  }
  %2 = util.scope "second part" {
    %t1 = stream.async.dispatch ...
    %t2 = stream.async.dispatch ...
    util.yield %t3
  }
}
```

**Comment by qedawkins** [comment]
> We could do this with an op interface so that we can have dialect-specific ops/syntactic sugar - like a hal.device.scope or something - but with external interfaces we could also make a single util.scope op implement dialect-specific interfaces.

I think it could be useful to have the op interface because we could combine it with other interfaces like having an `IsolatedFromAbove` version without needing to clone the op. What I'm thinking of is something like
```
func.func @my_kernel {
  %in_binding = ...
  %out_binding = ...
  %0 = codegen.scope.tu (%in_binding) {
    linalg.matmul
    ...
  } -> tensor<...>
  %1 = codegen.scope.tu (%0, %out_binding) {
    linalg.reduce
    ...
  }
}
```
The idea being, if I'm given a function to generate code for, I might want to use different strategies/pipelines for different portions of the function, but dialect conversions like bufferization will need to convert the whole function all at once (and might need similar operand tying to avoid extra allocations). Having an isolated scope lets us nest the pipeline on the scope without having to outline to a different function (and having to create another temporary module or something).

Not 100% sure this is the direction we'll want to go for codegen, but similar ideas have been thrown around a few times, so just food for thought.

**Comment by benvanik** [comment]
good point! my gut is that we'll want something else for codegen so it can implement interfaces that make sense to it (bufferization/etc) - we don't tend to use tied operands or shape-aware ops in there (just on host/prior to dispatch region formation). if such a codegen scope op existed earlier in the pipeline (prior to dispatch region formation) it'd probably be best to have it be distinct from whatever lower-level one would support memrefs/tiling/etc (if it even exists at that point) so that fusion/formation can handle it properly on tensors. I'd be interested to know if you could think of cases where we'd want a codegen-specific op early on that the flow/stream level needs to be aware of that isn't just setting the HAL target/etc? (gathering use cases)

**Comment by qedawkins** [comment]
> good point! my gut is that we'll want something else for codegen so it can implement interfaces that make sense to it (bufferization/etc) - we don't tend to use tied operands or shape-aware ops in there (just on host/prior to dispatch region formation). if such a codegen scope op existed earlier in the pipeline (prior to dispatch region formation) it'd probably be best to have it be distinct from whatever lower-level one would support memrefs/tiling/etc (if it even exists at that point) so that fusion/formation can handle it properly on tensors. I'd be interested to know if you could think of cases where we'd want a codegen-specific op early on that the flow/stream level needs to be aware of that isn't just setting the HAL target/etc? (gathering use cases)

Yeah something else for codegen is the key here, but I guess I'd need to see the interface for `util.scope` and/or work out the details to know whether this would be the exact abstraction we want there.

Re: codegen-specific scoping before dispatch formation, I can see a plausible use case for it, but I don't think we're close to supporting it. Thinking something along the lines of hand authored kernels while allowing for certain fusions. For example, I could replace a `linalg.matmul` with some hand authored (and pre-workgroup distributed code) that might want to fuse some elementwise stuff

```
func.func {
  %b = codegen.scope.workgroups {
    ... do some work
    codegen.store %some_tensor
  } -> some_tensor
  linalg.add %a, %b
}
```

There would be lots of trickiness in getting workgroup distribution to align, and potentially more trickiness when it comes to thread distribution on GPU (although that might just need more scoping) but I could see this as useful.

The other use case I was thinking of was scoping for custom kernel replacement, i.e. I match a dag and replace it with a `hal.dispatch.extern` or something. Initially my thought was we always jump straight to `hal.dispatch.extern`, but it might be useful to do matching even earlier on in, say, torch-mlir before tensor shapes have been resolved, and then use their shape propagation framework to update the region, then do outlining. Probably its own thing though if we go that far up.

Edit: Hmmm but thinking more about it, those are probably both different things. I don't see what Flow/Stream would need to know other than setting the target like you said.

**Comment by benvanik** [comment]
yeah, and it's always possible to just form your dispatch regions at that point instead. my current mental model where util.scope would be useful is "I want this layer of the model composed of dozens to hundreds of linalg ops to run on this device/with these assumptions/etc" and then have dispatch region formation run and produce one or more dispatches that are assigned to that device. if it's an individual dispatch/extern/etc then it's best just to form that dispatch directly.

**Comment by benvanik** [comment]
hmm not liking util.scope for user-level things - may add a util.annotate that is lighter-weight (no tied operand support, no dynamic dimension captures) that we can lower into util.scope for use within flow/stream/hal.
